kyaml/openapi/openapi.go

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package openapi
+
+import (
+	"sync"
+
+	"github.com/go-openapi/spec"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+var setup sync.Once
+var schema spec.Schema
+var schemaByResourceType map[yaml.TypeMeta]*spec.Schema
+
+// ResourceSchema wraps the OpenAPI Schema.
+type ResourceSchema struct {
+	// Schema is the OpenAPI schema for a Resource or field
+	Schema *spec.Schema
+}
+
+// SchemaForResourceType returns the Schema for the given Resource
+// TODO(pwittrock): create a version of this function that will return a schema
+// which can be used for duck-typed Resources -- e.g. contains common fields such
+// as metadata, replicas and spec.template.spec
+func SchemaForResourceType(t yaml.TypeMeta) *ResourceSchema {
+	initSchema()
+	rs, found := schemaByResourceType[t]
+	if !found {
+		return nil
+	}
+	return &ResourceSchema{Schema: rs}
+}
+
+// SchemaForElements returns the Schema for the elements of an array.
+func (r *ResourceSchema) SchemaForElements() *ResourceSchema {
+	// load the schema from swagger.json
+	initSchema()
+
+	if len(r.Schema.Type) != 1 || r.Schema.Type[0] != "array" {
+		// either not an array, or array has multiple types
+		return nil
+	}
+	s := *r.Schema.Items.Schema
+	for s.Ref.String() != "" {
+		sc, e := spec.ResolveRef(rootSchema(), &s.Ref)
+		if e != nil {
+			return nil
+		}
+		s = *sc
+	}
+	return &ResourceSchema{Schema: &s}
+}
+
+// SchemaForField returns the Schema for a field.
+func (r *ResourceSchema) SchemaForField(field string) *ResourceSchema {
+	// load the schema from swagger.json
+	initSchema()
+
+	// locate the Schema
+	s, found := r.Schema.Properties[field]
+	switch {
+	case found:
+		// no-op, continue with s as the schema
+	case r.Schema.AdditionalProperties != nil && r.Schema.AdditionalProperties.Schema != nil:
+		// map field type -- use Schema of the value
+		// (the key doesn't matter, they all have the same value type)
+		s = *r.Schema.AdditionalProperties.Schema
+	default:
+		// no Schema found from either swagger.json or line comments
+		return nil
+	}
+
+	// resolve the reference to the Schema if the Schema has one
+	for s.Ref.String() != "" {
+		sc, e := spec.ResolveRef(rootSchema(), &s.Ref)
+		if e != nil {
+			return nil
+		}
+		s = *sc
+	}
+
+	// return the merged Schema
+	return &ResourceSchema{Schema: &s}
+}
+
+// PatchStrategyAndKey returns the patch strategy and merge key extensions
+func (r *ResourceSchema) PatchStrategyAndKey() (string, string) {
+	ps, found := r.Schema.Extensions[kubernetesPatchStrategyExtensionKey]
+	if !found {
+		// merge key and patch strategy must appear together
+		return "", ""
+	}
+
+	mk, found := r.Schema.Extensions[kubernetesMergeKeyExtensionKey]
+	if !found {
+		// merge key and patch strategy must appear together
+		return "", ""
+	}
+	return ps.(string), mk.(string)
+}
+
+const (
+	// openAPIAssetName is the name of the asset containing the statically compiled in
+	// OpenAPI definitions for Kubernetes built-in types
```

Comment:
statistically compiled-in :grimacing:

Reply:
I think this is a good starting point.  Additional sources can be added later.

Reply:
All is well!